core/state/state_object.go

```diff
@@ -424,20 +424,21 @@ func (s *StateObject) updateTrie(db Database) Trie {
 		defer wg.Done()
 		// The snapshot storage map for the object
 		var storage map[string][]byte
+		if s.db.snap != nil {
+			s.db.snapMux.Lock()
+			if storage == nil {
+				// Retrieve the old storage map, if available, create a new one otherwise
+				if storage = s.db.snapStorage[s.address]; storage == nil {
+					storage = make(map[string][]byte)
+					s.db.snapStorage[s.address] = storage
+				}
+			}
+			s.db.snapMux.Unlock()
+		}
 		for key, value := range dirtyStorage {
 			// If state snapshotting is active, cache the data til commit
 			if s.db.snap != nil {
-				s.db.snapMux.Lock()
-				if storage == nil {
-					// Retrieve the old storage map, if available, create a new one otherwise
-					if storage = s.db.snapStorage[s.address]; storage == nil {
-						storage = make(map[string][]byte)
-						s.db.snapStorage[s.address] = storage
-					}
-				}
-
 				storage[string(key[:])] = value // value will be nil if value is 0x00
```

Comment:
`storage[string(key[:])] = value` this operation has to be protected by `s.db.snapMux.Lock()`
It is actually updating the `s.db.snapStorage`, it has concurrency issue.